query-connector/src/app/queryBuilding/components/ConditionSelection.tsx

```diff
@@ -0,0 +1,201 @@
+  >;
+  queryName: string;
+  validateForm: () => void;
+  setFormError: Dispatch<SetStateAction<FormError>>;
+  formError: FormError;
+  setLoading: Dispatch<SetStateAction<boolean>>;
+  loading: boolean;
+  setConditionValueSets: Dispatch<
+    SetStateAction<ConditionIdToValueSetArray | undefined>
+  >;
+};
+
+/**
+ * Display component for a condition on the query building page
+ * @param root0 - params
+ * @param root0.fetchedConditions - ID of the condition to reference
+ * @param root0.selectedConditions - name of condition to display
+ * @param root0.setBuildStep - Redirect function to handle view routing
+ * @param root0.setFetchedConditions - listener function for checkbox
+ * selection
+ * @param root0.setSelectedConditions - current checkbox selection status
+ * @param root0.queryName - current checkbox selection status
+ * @param root0.validateForm - function that checks for a valid query name
+ * and at least one selected condition
+ * @param root0.formError - indicates missing or incorrect form data
+ * @param root0.setFormError - state function that updates the status of the
+ * condition selection form input data
+ * @param root0.loading -  Boolean to track loading state
+ * @param root0.setLoading - The function to set the loading state
+ * @param root0.setConditionValueSets - state function that updates the status of the
+ * condition selection form input data
+ * @returns A component for display to redner on the query building page
+ */
+export const ConditionSelection: React.FC<ConditionSelectionProps> = ({
+  fetchedConditions,
+  selectedConditions,
+  setBuildStep,
+  setFetchedConditions,
+  setSelectedConditions,
+  queryName,
+  validateForm,
+  formError,
+  setFormError,
+  loading,
+  setConditionValueSets,
+  setLoading,
+}) => {
+  const focusRef = useRef<HTMLInputElement | null>(null);
+
+  const [searchFilter, setSearchFilter] = useState<string>();
+
+  useEffect(() => {
+    let isSubscribed = true;
+
+    if (queryName == "" || queryName == undefined) {
+      focusRef?.current?.focus();
+    }
+
+    async function fetchConditionsAndUpdateState() {
+      const { categoryToConditionArrayMap } = await getConditionsData();
+
+      if (isSubscribed) {
+        setFetchedConditions(
+          mapFetchedDataToFrontendStructure(categoryToConditionArrayMap),
+        );
+      }
+    }
+
+    fetchConditionsAndUpdateState().catch(console.error);
+    return () => {
+      isSubscribed = false;
+    };
+  }, []);
+
+  async function getValueSetsForSelectedConditions() {
+    const conditionIds = Object.entries(selectedConditions)
+      .map(([_, conditionObj]) => {
+        return Object.keys(conditionObj);
+      })
+      .flatMap((ids) => ids);
+
+    const ConditionValueSets: ConditionIdToValueSetArray = {};
+
+    for (const id of conditionIds) {
+      const results: ValueSet[] =
+        await getValueSetsAndConceptsByConditionID(id);
+
+      const formattedResults = await mapQueryRowsToValueSets(results);
```

Comment:
This might be a good opportunity for us to pull out `mapQueryRowsToValueSets` from the `database-service.ts` and make it a synchronous function. If you read the function code, there's no actual promise that's getting awaited that necessitates it being an async function.

I think the only reason that function is marked as async is because that file is a server action that requires any exported function to be marked as async. I think if we moved it into a utils file that isn't a server actions file, we can make this not be async true to the nature of the function.